include/lapack/gelq2.hpp

```diff
@@ -60,24 +61,37 @@ namespace tlapack
         using idx_t = size_type<matrix_t>;
         using range = std::pair<idx_t, idx_t>;
 
+        // constants
         const idx_t m = nrows(A);
         const idx_t n = ncols(A);
         const idx_t k = std::min(m, n);
 
+        // check arguments
+        tlapack_check_false(access_denied(dense, write_policy(A)), -1);
+        tlapack_check_false((idx_t)size(tauw) < std::min<idx_t>(m, n), -2);
+        tlapack_check_false((idx_t)size(work) < m, -3);
+
+        // quick return
+        if (m <= 0)
+            return 0;
+
         for (idx_t j = 0; j < k; ++j)
         {
+            // Define w := A(j,j:n)
             auto w = slice(A, j, range(j, n));
+
+            // Generate elementary reflector H(j) to annihilate A(j,j+1:n)
             for (idx_t i = 0; i < n - j; ++i)
                 w[i] = conj(w[i]); // see LAPACK about conjugating w
-
-            larfg(w, tauw[j]); // generate the horizontal reflector w
+            larfg(w, tauw[j]);
 
             if (j < k - 1 || k < m)
             {
+                // Apply H(j) to A(j+1:m,j:n) from the right
                 auto Q11 = slice(A, range(j + 1, m), range(j, n));
                 larf(Side::Right, w, tauw[j], Q11, work);
             }
-            for (idx_t i = 0; i < n - j; ++i) 
+            for (idx_t i = 0; i < n - j; ++i)
```

Comment:
@langou Hey Julien, so the algorithms for GEBD2 and GELQ2 are similar here, they conjugate `w` before and after `larfg` and `larf`. In GELQ2 (and the UNGL2 `else` part to pair with it) I just used the LAPACK ZGELQ2 here. 

So Weslley found that we can just conjugate `w` for once before `larfg` in GEBD2 (and no conj(w) in UNGL2 `if (bidg == 1) ` part, but with `conj(tauw)` in `larf),` and this is pretty much the same for GELQ2 (in this line 94 `for` loop can go away along with the two `conj(w)`s in UNGL2). 

You mentioned you want me to put the bidiagonal option in GELQ2 but we don't really use it here. Do you still want me to include the bidiagonal in it? Would you want me to change the `conj(w)` algorithms for GELQ2 and UNGL2 (as GEBD2) as well?

Reply:
I decided that I'll go two different routines for UNGL2 and generating matrix Z for GEBD2, since they are indeed doing different things. Therefore, I'll do another .hpp file for generating Z and remove the bidiagonal boolean in UNGL2 to make it cleaner.

Reply:
I think @thijssteel and I agree with splitting into two routines. Nice job @yuxincai11!